fix: Validation cannot handle array item

kenjis opened the pull request:
**Description**
Supersedes #5402
Fixes #5368
- separate validation process of multiple field and single field

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

MGatner reviewed (changes requested):
The code is a bit hard to review on mobile but what I see all looks good, and the tests are great. 

Your User Guide write-up is very thorough, but I would rather see that in the changelogs (i.e. ** user_guide_src/source/changelogs/v4.1.6.rst**) and keep the Upgrading section as actual *tasks* that need to be done when upgrading. A simple sentence with reference to the longer explanation in the changelog would be good, like "Due to a bug fix Validation now might ____ (see ____.rst), so make sure you ____ to any code calling it."

MGatner commented:
Will this conflict with the Strict Validation changes?

kenjis commented:
@MGatner I added the last commit to fix the docs. I rebased because the commits are not verified.

> Will this conflict with the Strict Validation changes?

No.
#5402 will be split into two PRs. This is one.

MGatner reviewed (approved):
Doc changes look good! Since this is a breaking change I'd like it if you could get another reviewer. I'll ping the team on Slack.

paulbalandan commented on system/Validation/Validation.php:
```diff
@@ -139,16 +139,20 @@ public function run(?array $data = null, ?string $group = null, ?string $dbGroup
+                }
+            } else {
+                // Process single field
+                $this->processRules($field, $setup['label'] ?? $field, $values, $rules, $data);
```
Since this one is the same with L145, I think we can combine this with that?

kenjis replied:
If I remove the `if` statement from L143, tests in `tests/system/Validaton/` passes.
But I'm not sure removing it causes no problem.

What input makes `$values` empty array?

kenjis replied:
I added `continue` in the `if` to fix the bug that `$this->processRules()` run twice.

paulbalandan replied:
> What input makes `$values` empty array?

This is raised in an issue report. I'll check.

paulbalandan replied:
https://github.com/codeigniter4/CodeIgniter4/issues/4510#issuecomment-813532899

kenjis replied:
I read the comment and the fix PR.
But I'm not sure we can remove the `if ($values === []) { ... }`.
I feel it is safer not to remove it.

kenjis replied:
@paulbalandan I want to update `upgrade_416.rst`, so I merge this for now.

paulbalandan commented on system/Validation/Validation.php:
```diff
@@ -274,7 +285,8 @@ protected function processRules(string $field, ?string $label, $value, $rules =
 
                     $found  = true;
-                    $passed = $param === false ? $set->{$rule}($value, $error) : $set->{$rule}($value, $param, $data, $error);
+                    $passed = $param === false ? $set->{$rule}($value, $error)
+                        : $set->{$rule}($value, $param, $data, $error);
```
The `operator_linebreak` should break this one.
```suggestion
                    $passed = $param === false
                        ? $set->{$rule}($value, $error)
                        : $set->{$rule}($value, $param, $data, $error);
```

paulbalandan commented on system/Validation/Validation.php:
```diff
@@ -295,7 +307,14 @@ protected function processRules(string $field, ?string $label, $value, $rules =
                 }
 
-                $this->errors[$field] = $error ?? $this->getErrorMessage($rule, $field, $label, $param, $value);
+                $param                = ($param === false) ? '' : $param;
+                $this->errors[$field] = $error ?? $this->getErrorMessage(
```
Can we add a line in between? The empty space between `$param` and `=` is quite much.

kenjis replied:
Done.

paulbalandan commented on user_guide_src/source/changelogs/v4.1.6.rst:
```diff
@@ -7,12 +7,26 @@ Release Date: Not released
+------------------
+- The previous version of the Validation can't handle an array item.
+  Because of the bug fix, the validation results may be different,
+  or raise an TypeError.
```
```suggestion
  or raise a ``TypeError``.
```

samsonasik commented on system/Validation/Validation.php:
```diff
@@ -646,6 +669,9 @@ public function setError(string $field, string $error): ValidationInterface
      */
     protected function getErrorMessage(string $rule, string $field, ?string $label = null, ?string $param = null, ?string $value = null): string
     {
+        $param = $param ?? '';
```
how about `(string) $param` ? ref https://3v4l.org/NijdY#v7.3.33

kenjis replied:
@samsonasik What is better?

samsonasik replied:
probably just cast it? 

```suggestion
        $param = (string) $param;
```

kenjis replied:
`$param` is null or string. So `$param = $param ?? ''` is no problem.

I'm not sure, but it seems @paulbalandan likes `$param = $param ?? ''`.
Is there any advantage of `(string)`?

samsonasik replied:
oh, ok it seems other file have it as well, ok then 👍

paulbalandan replied:
I don't know of optimization gains but typing `?? ''` is much faster than `(string)` 😂
